### PR TITLE
WGAN in advent (This PR also includes the former compare loss which is now GroundIntersectionLoss)

### PR DESCRIPTION
**Pull request by tianyu-z**
I added an option for ADVENT discriminator:

dis.m.gan_type now can be "GAN", "WGAN", "WGAN_gp", "WGAN_norm" and I added a "rmsprop" option for our optimizer which works for "WGAN" (Only "WGAN", not  "WGAN_gp" and "WGAN_norm").

### CODE REVIEW

File: omnigan/trainer.py [hunk 1]
```diff
@@ -836,6 +836,7 @@ def get_masker_loss(self, multi_domain_batch):  # TODO update docstrings
                                     prediction,
                                     self.domain_labels["s"],
                                     self.D["s"]["Advent"],
+                                    gan_type=self.opts.dis.s.gan_type,
```
Comment (vict0rsch) [inline omnigan/trainer.py:839]: This cannot work, gan_type should be provided at init time not call time

**Review by vict0rsch — CHANGES_REQUESTED** [review]

File: omnigan/trainer.py (hunk 1 above)
Comment (tianyu-z) [reply omnigan/trainer.py:839]: Yes. I gave up testing because of some network issues last night. I am about to fix that.

**Comment by tianyu-z** [comment]
Now, seg advent and masker seg can call different kinds of GAN_type by dis.m.gan_type and dis.s.gan_type

**Comment by tianyu-z** [comment]
Training exps:
WGAN with mask only:
https://www.comet.ml/tianyu-z/omnigan/01f839bc294a4e569e7f528a1a7017b0?experiment-tab=chart&showOutliers=true&smoothing=0&transformY=smoothing&xAxis=step

WGAN with m+s:
https://www.comet.ml/tianyu-z/omnigan/a19d3e666c004573804353a562457973?experiment-tab=chart&showOutliers=true&smoothing=0&transformY=smoothing&xAxis=step

WGAN_gp with mask only:
https://www.comet.ml/tianyu-z/omnigan/efacf54c359d4cd1ba21e8df52065b73?experiment-tab=chart&showOutliers=true&smoothing=0&transformY=smoothing&xAxis=step

WGAN_gp with m+s:
https://www.comet.ml/tianyu-z/omnigan/bfa8d1b4ff444e7b93a7ddddb7fd756b?experiment-tab=chart&showOutliers=true&smoothing=0&transformY=smoothing&xAxis=step

WGAN_norm with mask only:
https://www.comet.ml/tianyu-z/omnigan/1c034e4d64f94cba912e028aa16d0c0e?experiment-tab=chart&showOutliers=true&smoothing=0&transformY=smoothing&xAxis=step

WGAN_norm with m+s:
https://www.comet.ml/tianyu-z/omnigan/c81c59dd659343deac7f30de6f7384b9?experiment-tab=chart&showOutliers=true&smoothing=0&transformY=smoothing&xAxis=step

File: omnigan/trainer.py [hunk 2]
```diff
@@ -871,6 +872,21 @@ def get_masker_loss(self, multi_domain_batch):  # TODO update docstrings
...
 
+                    # Then GroundIntersection loss
+                    if self.opts.gen.m.use_ground_intersection:
+                        if self.verbose > 0:
+                            print("Using GroundIntersection loss.")
+                        update_loss = (
+                            self.losses["G"]["tasks"][update_task]["gi"](
+                                prediction, update_target
+                            )
+                            * lambdas.G[update_task]["gi"]
```
Comment (vict0rsch) [inline omnigan/trainer.py:883]: I think we should only use GI on the real world otherwise this signal will conflict with the actual ground truth on slopes in the simulated world (as you know, ground != physical horizontal)

File: omnigan/trainer.py [hunk 3]
```diff
@@ -1179,10 +1195,39 @@ def get_d_loss(self, multi_domain_batch, verbose=0):
...
+                                )
+                            disc_loss["m"]["Advent"] += (
+                                self.opts.train.lambdas.advent.adv_main * loss_main
+                            )
+                        elif self.opts.dis.m.gan_type == "WGAN_gp":
+                            prob_need_grad = autograd.Variable(prob, requires_grad=True)
+                            d_out = self.D["m"]["Advent"](prob_need_grad)
+                            # github code reference: https://github.com/caogang/wgan-gp/blob/master/gan_cifar10.py
+                            grads = autograd.grad(
+                                outputs=d_out,
```
Comment (vict0rsch) [inline omnigan/trainer.py:1216]: Can you move that to a `tutils` function to keep this loop as clean as possible?

File: omnigan/trainer.py [hunk 4]
```diff
@@ -1194,9 +1239,39 @@ def get_d_loss(self, multi_domain_batch, verbose=0):
...
+                                )
+                            disc_loss["s"]["Advent"] += (
+                                self.opts.train.lambdas.advent.adv_main * loss_main
+                            )
+                        elif self.opts.dis.s.gan_type == "WGAN_gp":
+                            prob_need_grad = autograd.Variable(prob, requires_grad=True)
+                            d_out = self.D["s"]["Advent"](prob_need_grad)
+                            # github code reference: https://github.com/caogang/wgan-gp/blob/master/gan_cifar10.py
+                            grads = autograd.grad(
+                                outputs=d_out,
```
Comment (vict0rsch) [inline omnigan/trainer.py:1260]: same here

**Review by vict0rsch — CHANGES_REQUESTED** [review]

**Comment by tianyu-z** [comment]
@vict0rsch Fixed

**Comment by vict0rsch** [comment]
I agree with you masking alone with WGAN+spectral norm is the best. We need to make the segmentation work! :)

**Review by vict0rsch — APPROVED** [review]

**Review by melisandeteng — APPROVED** [review]
